### PR TITLE
ceph: added monitoring managedBy label to metrics

### PR DESCRIPTION
**Pull request by anmolsachan**
This PR depends on  https://github.com/openshift/ocs-operator/pull/1276.

This PR adds 'managedBy' label to the ceph metrics if the
cephcluster is managed by another higher Resource. The optional label shall
be added as part of cephcluster CRs monitoring labels.

Signed-off-by: Anmol Sachan <anmol13694@gmail.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [x] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [x] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.
- [x] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [x] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/cluster/mgr/mgr.go [hunk 1]
```diff
@@ -467,6 +468,28 @@ func (c *Cluster) EnableServiceMonitor(activeDaemon string) error {
 	}
 	serviceMonitor.Spec.NamespaceSelector.MatchNames = []string{c.clusterInfo.Namespace}
 	serviceMonitor.Spec.Selector.MatchLabels = c.selectorLabels(activeDaemon)
+
+	// If the cephcluster is managed by another resource,
+	// add the name of that resource as a label on the ceph metrics.
+	if c.spec.Labels != nil {
+		monitoringLabels, ok := c.spec.Labels["monitoring"]
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/mgr.go:475]: nit on consolidating the check for `ok`:
```suggestion
		if monitoringLabels, ok := c.spec.Labels["monitoring"]; ok {
```

File: pkg/operator/ceph/cluster/mgr/mgr.go [hunk 2]
```diff
@@ -467,6 +468,28 @@ func (c *Cluster) EnableServiceMonitor(activeDaemon string) error {
...
+			managedBy, ok := monitoringLabels["managedBy"]
+			if ok {
+				relabelConfig := monitoringv1.RelabelConfig{
+					TargetLabel: "managedBy",
+					Replacement: managedBy,
+				}
+				serviceMonitor.Spec.Endpoints[0].RelabelConfigs = append(
+					serviceMonitor.Spec.Endpoints[0].RelabelConfigs, &relabelConfig)
+			} else {
+				logger.Errorf("Could not find managedBy label in Monitoring labels.")
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/mgr.go:486]: These labels are completely optional, right? If so, let's just print info.
```suggestion
				logger.Info("Could not find managedBy label in Monitoring labels.")
```

File: pkg/operator/ceph/cluster/mgr/mgr.go [hunk 3]
```diff
@@ -467,6 +468,28 @@ func (c *Cluster) EnableServiceMonitor(activeDaemon string) error {
...
+			}
+		} else {
+			logger.Errorf("Could not find Monitoring labels.")
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/mgr.go:489]:
```suggestion
			logger.Info("Could not find Monitoring labels.")
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/cluster/mgr/mgr.go (hunk 1 above)
Comment (anmolsachan) [reply pkg/operator/ceph/cluster/mgr/mgr.go:475]: Ack. I will update.

File: pkg/operator/ceph/cluster/mgr/mgr.go (hunk 2 above)
Comment (anmolsachan) [reply pkg/operator/ceph/cluster/mgr/mgr.go:486]: Yes, the labels are optional.

File: pkg/operator/ceph/cluster/mgr/mgr.go [hunk 4]
```diff
@@ -467,6 +468,26 @@ func (c *Cluster) EnableServiceMonitor(activeDaemon string) error {
...
+	// If the cephcluster is managed by another resource,
+	// add the name of that resource as a label on the ceph metrics.
+	if c.spec.Labels != nil {
+		if monitoringLabels, ok := c.spec.Labels["monitoring"]; ok {
+			if managedBy, ok := monitoringLabels["rook.io/managedBy"]; ok {
+				relabelConfig := monitoringv1.RelabelConfig{
+					TargetLabel: "managedBy",
+					Replacement: managedBy,
+				}
+				serviceMonitor.Spec.Endpoints[0].RelabelConfigs = append(
```
Comment (sp98) [inline pkg/operator/ceph/cluster/mgr/mgr.go:481]: won't `RelabelConfigs` need a `nil` check?

**Review by sp98 — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/cluster/mgr/mgr.go (hunk 4 above)
Comment (anmolsachan) [reply pkg/operator/ceph/cluster/mgr/mgr.go:481]: @sp98 Can you please elaborate on that?

File: pkg/operator/ceph/cluster/mgr/mgr.go [hunk 5]
```diff
@@ -467,6 +468,26 @@ func (c *Cluster) EnableServiceMonitor(activeDaemon string) error {
...
+					serviceMonitor.Spec.Endpoints[0].RelabelConfigs, &relabelConfig)
+			} else {
+				logger.Info("Could not find rook.io/managedBy label in Monitoring labels.")
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/mgr.go:484]:
```suggestion
				logger.Info("rook.io/managedBy not specified in monitoring labels")
```

File: pkg/operator/ceph/cluster/mgr/mgr.go [hunk 6]
```diff
@@ -467,6 +468,26 @@ func (c *Cluster) EnableServiceMonitor(activeDaemon string) error {
...
+			}
+		} else {
+			logger.Info("Could not find Monitoring labels.")
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/mgr.go:487]:
```suggestion
			logger.Info("monitoring labels not specified")
```

**Review by travisn — CHANGES_REQUESTED** [review]
just nits on the log messages

File: pkg/operator/ceph/cluster/mgr/mgr.go (hunk 4 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/mgr/mgr.go:481]: appending to a nil slice would actually initialize the slice, so it looks good

**Review by travisn — APPROVED** [review]

**Review by sp98 — APPROVED** [review]

File: pkg/operator/ceph/cluster/mgr/mgr.go [hunk 7]
```diff
@@ -467,6 +468,26 @@ func (c *Cluster) EnableServiceMonitor(activeDaemon string) error {
 	}
 	serviceMonitor.Spec.NamespaceSelector.MatchNames = []string{c.clusterInfo.Namespace}
 	serviceMonitor.Spec.Selector.MatchLabels = c.selectorLabels(activeDaemon)
+
+	// If the cephcluster is managed by another resource,
+	// add the name of that resource as a label on the ceph metrics.
+	if c.spec.Labels != nil {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/mgr.go:474]: Could you actually factor out this new code block into a method such as `applyMonitoringLabels()`, then add a unit test for that new method?

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/cluster/mgr/mgr.go (hunk 7 above)
Comment (anmolsachan) [reply pkg/operator/ceph/cluster/mgr/mgr.go:474]: Ack, Sure I will do that.

Comment (anmolsachan) [reply pkg/operator/ceph/cluster/mgr/mgr.go]: @travisn I updated the PR according to your recommendation. Please review. Thank you.

File: pkg/operator/ceph/cluster/mgr/mgr.go [hunk 8]
```diff
@@ -506,3 +510,24 @@ func IsModuleInSpec(modules []cephv1.Module, moduleName string) bool {
 
 	return false
 }
+
+// ApplyMonitoringLabels function adds the name of the resource that manages
+// cephcluster, as a label on the ceph metrics
+func ApplyMonitoringLabels(c *Cluster, serviceMonitor *monitoringv1.ServiceMonitor) {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/mgr.go:516]: thanks for refactoring, just one final nit:
```suggestion
func applyMonitoringLabels(c *Cluster, serviceMonitor *monitoringv1.ServiceMonitor) {
```

Comment (anmolsachan) [reply pkg/operator/ceph/cluster/mgr/mgr.go]: Thank you for the review. I have updated it according to the recomnendation.

**Review by travisn — APPROVED** [review]

**Pull request merged**